src/precice/impl/SolverInterfaceImpl.cpp

```diff
@@ -1564,22 +1665,21 @@ void SolverInterfaceImpl::mapData(const utils::ptr_vector<DataContext> &contexts
   PRECICE_TRACE();
   using namespace mapping;
   MappingConfiguration::Timing timing;
+  int                          sampleID = 0;
   for (impl::DataContext &context : contexts) {
     if (context.hasMapping()) {
       timing         = context.mappingContext().timing;
       bool hasMapped = context.mappingContext().hasMappedData;
       bool mapNow    = timing == MappingConfiguration::ON_ADVANCE;
       mapNow |= timing == MappingConfiguration::INITIAL;
       if (mapNow && (not hasMapped)) {
-        int inDataID  = context.getFromDataID();
-        int outDataID = context.getToDataID();
         PRECICE_DEBUG("Map \"{}\" data \"{}\" from mesh \"{}\"",
                       mappingType, context.getDataName(), context.getMeshName());
         context.resetToData();
-        PRECICE_DEBUG("Map from dataID {} to dataID: {}", inDataID, outDataID);
-        context.mappingContext().mapping->map(inDataID, outDataID);
-        PRECICE_DEBUG("Mapped values = {}", utils::previewRange(3, context.toData()->values())); // @todo might be better to move this debug message into Mapping::map and remove getter DataContext::toData()
+        context.mapWaveformSample(sampleID);
```

Comment:
Main reason for this change is that I want to avoid long chains of calls. This makes it harder to understand who accesses what. I think this also allows me to not expose the `mappingContext` to the `SolverInterfaceImpl` at all and let the `dataContext` handle everything.

Reply:
`DataContext`/`ReadDataContext` now provide one function for each one of the relevant mappings:

* `DataContext::mapWrittenData` is called by `SolverInterfaceImpl::mapWrittenData`
* `DataContext::mapWriteDataFrom` is called by `SolverInterfaceImpl::mapWriteDataFrom`
* `ReadDataContext::mapReadData` is called by `SolverInterfaceImpl::mapReadData`
* `ReadDataContext::mapReadDataTo` is called by `SolverInterfaceImpl::mapReadDataTo`

Generally these functions can also be moved directly into `SolverInterfaceImpl`, but I prefer to hide as much as possible inside the `DataContext` and reduce complexity on the side of `SolverInterfaceImpl`.

Reply:
Mmh, I must say I don't really agree. I think it makes it easier to have this explicitly in `SolverInterfaceImpl` and not hidden. In the end, mapping is one of the core features and it should be easy to spot, from where it is called. 
If you put the mapping into `ReadDataContext`, one could also argue to put the communication and the acceleration in there. No?

Reply:
Moved the mapping back into `SolverInterfaceImpl`